langchain/document_loaders/text.py

```diff
@@ -1,20 +1,55 @@
-from typing import List, Optional
+from typing import TYPE_CHECKING, List, Optional
+
+import chardet
 
 from langchain.docstore.document import Document
 from langchain.document_loaders.base import BaseLoader
 
+if TYPE_CHECKING:
+    from chardet.resultdict import ResultDict
+
 
 class TextLoader(BaseLoader):
     """Load text files."""
 
-    def __init__(self, file_path: str, encoding: Optional[str] = None):
+    def __init__(self,
+                 file_path: str,
+                 encoding: Optional[str] = None,
+                 autodetect_encoding: bool = False):
         """Initialize with file path."""
         self.file_path = file_path
         self.encoding = encoding
+        self.autodetect_encoding = autodetect_encoding
 
     def load(self) -> List[Document]:
         """Load from file path."""
         with open(self.file_path, encoding=self.encoding) as f:
-            text = f.read()
+            try:
+                text = f.read()
+            except UnicodeDecodeError:
+                if self.autodetect_encoding:
+                    detected_encodings = self.detect_file_encodings()
+                    for encoding in detected_encodings:
+                        try:
+                            with open(self.file_path,
+                                      encoding=encoding['encoding']) as f:
+                                text = f.read()
+                            break
+                        except UnicodeDecodeError:
+                            continue
+                else:
+                    raise
+            except Exception as e:
+                raise RuntimeError(f"Error loading {self.file_path}") from e
+
         metadata = {"source": self.file_path}
         return [Document(page_content=text, metadata=metadata)]
+
+    def detect_file_encodings(self) -> List['ResultDict']:
+        """Try to detect the file encoding."""
+        with open(self.file_path, 'rb') as f:
+            rawdata = f.read()
+        return [
+            encoding for encoding in chardet.detect_all(rawdata)
```

Comment:
nit: could we make it `[encoding['encoding'] for encoding in ...]`

Reply:
I had to make more changes here to make sure to fail if no encoding is detected